package com.Proyecto.coffeepalace.Data.Remote.auth.Register

import com.Proyecto.coffeepalace.Data.Remote.auth.Login.LoginRequest
import com.Proyecto.coffeepalace.Data.Remote.auth.Login.LoginResponse
import com.Proyecto.coffeepalace.Data.Remote.auth.RegisterRequest
import com.Proyecto.coffeepalace.Data.Remote.auth.UserResponse
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Part
import retrofit2.http.Path

interface AuthService {
    @POST("users/login")
    suspend fun login(@Body request: LoginRequest): LoginResponse

    @POST("users/")
    suspend fun register(@Body request: RegisterRequest): LoginResponse
    @Multipart
    @POST("users/")
    suspend fun registerMultipart(
        @Part("forenames") forenames: RequestBody,
        @Part("surnames") surnames: RequestBody,
        @Part("email") email: RequestBody,
        @Part("password") password: RequestBody,
        @Part("phone_number") phoneNumber: RequestBody,
        @Part("gender") gender: RequestBody,
        @Part("role_id") roleId: RequestBody,
        @Part profile_pic: MultipartBody.Part?
    ): LoginResponse

    @POST("users/logout")
    suspend fun logout(): Unit
    @GET("users/")
    suspend fun getAllUsers(@Header("Authorization") token: String): List<UserResponse>
    @GET("register-codes/users")
    suspend fun getAllUsersWithCodes(@Header("Authorization") token: String): List<UserResponse>
    @Multipart
    @PUT("users/{id}")
    suspend fun updateUser(
        @Path("id") userId: Int,
        @Part("forenames") forenames: RequestBody,
        @Part("surnames") surnames: RequestBody,
        @Part("email") email: RequestBody,
        @Part("phone_number") phone_number: RequestBody,
        @Part("gender") gender: RequestBody,
        @Part profile_pic: MultipartBody.Part?,
        @Header("Authorization") authHeader: String
    ): UserResponse

    @PUT("users/{id}/password")
    suspend fun changePassword(
        @Path("id") userId: Int,
        @Body body: Map<String, String>,
        @Header("Authorization") authHeader: String
    ): retrofit2.Response<okhttp3.ResponseBody>

    @DELETE("users/{id}")
    suspend fun deleteAccount(
        @Path("id") userId: Int,
        @Header("Authorization") authHeader: String
    ): retrofit2.Response<okhttp3.ResponseBody>
    @POST("users/recover-password")
    suspend fun recoverPassword(@Body body: Map<String, String>)
    @POST("users/verify-reset-code")
    suspend fun verifyResetCode(@Body body: Map<String, String>)
    @GET("users/{id}")
    suspend fun getUserById(
        @Path("id") userId: Int,
        @Header("Authorization") authHeader: String
    ): UserResponse
}